src/NanopassSharp/ITransformation.cs

```diff
@@ -0,0 +1,38 @@
+﻿namespace NanopassSharp;
+
+/// <summary>
+/// A transformation on a node tree.
+/// </summary>
+public interface ITransformation
+{
+
+    /// <summary>
+    /// The pattern the transformation is triggered by.
+    /// </summary>
+    ITransformationPattern? Pattern { get; }
```

Comment:
I wonder if we should decouple the pattern and the transformation. It could reduce code-duplication, as the transformation would not need to implement querying for relevant nodes each time. In exchange, `Apply` would change to something like `Apply(Tree, Node)`.

Reply:
So a transformation would only handle the transformation of individual types as opposed to entire tree? How would the transformation know what to transform if not supplied with a pattern to work off of?

Reply:
Yes, it would handle the individual node. I'd say the situation depends on the transformation. A transformation might not need a pattern because it works with the entire tree (in which case the node would be null). If a pattern is not provided for a transformation that works on some node, you could come up with a sane default (like all nodes are transformed or something).

You could also keep the current design, I just feel like the node matching logic should be externalized to minimize code repetition.

Reply:
> I just feel like the node matching logic should be externalized to minimize code repetition.

I agree with this, although I'm still wondering about the transformation stuff. If a transformation should transform a member of a node, how would it know which member to transform without having a pattern to find the member(s) by? Or are members also "nodes"?

Reply:
Ooooh fair point, I forgot about members honestly. That's a tough one, I have no immediate answer to that.

Reply:
I could personally see two possible solutions for this:

1. `ITransformation` has two overloads for `Apply`, `Apply(Tree, TypeNode)` and `Apply(Tree, TypeNode, MemberNode)`. For transformations which only transform one of them, the other can simply return the input tree. 

2. Alternatively it could be `Apply(Tree, TypeNode, MemberNode?)` where the member could be null if only the type should be transformed.

3. There are two sub-interfaces of `ITransformation`, `ITypeTransformation` which deals with transformations of types, and `IMemberTransformation` which deals with transformations of members. A simple switch expression would be enough to call the appropriate method.